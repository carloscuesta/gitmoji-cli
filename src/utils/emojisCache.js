// @flow
import fs from 'fs'
import os from 'os'
import path from 'path'
import pathExists from 'path-exists'

export const GITMOJI_CACHE: Object = {
  FOLDER: '.gitmoji',
  FILE: 'gitmojis.json'
}

export const CACHE_PATH = path.join(
  os.homedir(),
  GITMOJI_CACHE.FOLDER,
  GITMOJI_CACHE.FILE
)

const createEmojis = (emojis: Array<Object>) => {
  if (!pathExists.sync(path.dirname(CACHE_PATH))) {
    fs.mkdirSync(path.dirname(CACHE_PATH))
  }

  fs.writeFileSync(CACHE_PATH, JSON.stringify(emojis))
}

const getEmojis = () => {
  // $FlowFixMe
  return Promise.resolve(JSON.parse(fs.readFileSync(CACHE_PATH)))
}

const isAvailable = () => pathExists.sync(CACHE_PATH)

export default {
  createEmojis,
  getEmojis,
  isAvailable
}
