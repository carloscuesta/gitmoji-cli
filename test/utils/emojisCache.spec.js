import pathExists from 'path-exists'
import fs from 'fs'
import path from 'path'

import emojisCache, {
  GITMOJI_CACHE,
  CACHE_PATH
} from '../../src/utils/emojisCache'
import * as stubs from './stubs'

jest.mock('os', () => ({
  homedir: jest.fn().mockReturnValue('/Users/Test/')
}))

describe('emojisCache', () => {
  it('should match GITMOJI_CACHE', () => {
    expect(GITMOJI_CACHE).toMatchSnapshot()
  })

  it('should match CACHE_PATH', () => {
    expect(CACHE_PATH).toMatchSnapshot()
  })

  describe('isAvailable', () => {
    emojisCache.isAvailable()

    it('should check that cache path exists', () => {
      expect(pathExists.sync).toHaveBeenCalledWith(CACHE_PATH)
    })
  })

  describe('createEmojis', () => {
    describe('when cache directory does not exists', () => {
      beforeAll(() => {
        pathExists.sync.mockReturnValue(false)
        emojisCache.createEmojis(stubs.gitmojis)
      })

      it('should create cache directory', () => {
        expect(fs.mkdirSync).toHaveBeenCalledWith(path.dirname(CACHE_PATH))
      })

      it('should write the gitmojis.json file', () => {
        expect(fs.writeFileSync).toHaveBeenCalledWith(
          CACHE_PATH,
          JSON.stringify(stubs.gitmojis)
        )
      })
    })

    describe('when cache directory exists', () => {
      beforeAll(() => {
        pathExists.sync.mockReturnValue(true)
        fs.mkdirSync.mockClear()
        fs.writeFileSync.mockClear()
        emojisCache.createEmojis(stubs.gitmojis)
      })

      it('should create cache directory', () => {
        expect(fs.mkdirSync).not.toHaveBeenCalled()
      })

      it('should write the gitmojis.json file', () => {
        expect(fs.writeFileSync).toHaveBeenCalledWith(
          CACHE_PATH,
          JSON.stringify(stubs.gitmojis)
        )
      })
    })
  })

  describe('getEmojis', () => {
    beforeAll(() => {
      fs.readFileSync.mockReturnValue(JSON.stringify(stubs.gitmojis))
      emojisCache.getEmojis()
    })

    it('should read the emojis from the cache', () => {
      expect(fs.readFileSync).toHaveBeenCalled()
    })
  })
})
