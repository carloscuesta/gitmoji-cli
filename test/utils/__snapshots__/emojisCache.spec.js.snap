// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emojisCache should match CACHE_PATH 1`] = `"/Users/Test/.gitmoji/gitmojis.json"`;

exports[`emojisCache should match GITMOJI_CACHE 1`] = `
Object {
  "FILE": "gitmojis.json",
  "FOLDER": ".gitmoji",
}
`;
